algorithms/unit_tests/TestSort.hpp

```diff
@@ -402,6 +403,339 @@ void test_sort_integer_overflow() {
+    }
+  }
+  KeyViewType keys;
+  OffsetViewType offsets;
+  bool sortDescending;
+};
+
+// Functor to test sort_by_key_thread
+template <typename ExecSpace, typename KeyViewType, typename ValueViewType,
+          typename OffsetViewType>
+struct ThreadSortByKeyFunctor {
+  using TeamMem  = typename Kokkos::TeamPolicy<ExecSpace>::member_type;
+  using SizeType = typename KeyViewType::size_type;
+  using KeyType  = typename KeyViewType::non_const_value_type;
+  ThreadSortByKeyFunctor(const KeyViewType& keys_, const ValueViewType& values_,
+                         const OffsetViewType& offsets_, bool sortDescending_)
+      : keys(keys_),
+        values(values_),
+        offsets(offsets_),
+        sortDescending(sortDescending_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const TeamMem& t) const {
+    int i = t.league_rank() * t.team_size() + t.team_rank();
+    // Number of arrays to sort doesn't have to be divisible by team size, so
+    // some threads may be idle.
+    if (i < offsets.extent_int(0) - 1) {
+      SizeType begin = offsets(i);
+      SizeType end   = offsets(i + 1);
+      if (sortDescending) {
+        Kokkos::Experimental::sort_by_key_thread(
+            t, Kokkos::subview(keys, Kokkos::make_pair(begin, end)),
+            Kokkos::subview(values, Kokkos::make_pair(begin, end)),
+            GreaterThan<KeyType>());
+      } else {
+        Kokkos::Experimental::sort_by_key_thread(
+            t, Kokkos::subview(keys, Kokkos::make_pair(begin, end)),
+            Kokkos::subview(values, Kokkos::make_pair(begin, end)));
+      }
+    }
+  }
+  KeyViewType keys;
+  ValueViewType values;
+  OffsetViewType offsets;
+  bool sortDescending;
+};
+
+// Randomly generate n arrays of up to k elements each, packed into a single 1D
+// view. Like a row in a StaticCrsGraph, array i begins at offsets(i)
+// (inclusive) and ends at offsets(i+1) (exclusive)
+template <typename KeyType, typename KeyViewType, typename OffsetViewType>
+void randomPackedArrays(unsigned n, unsigned k, KeyViewType& keys,
+                        OffsetViewType& offsets, KeyType minKey,
+                        KeyType maxKey) {
+  offsets          = OffsetViewType("Offsets", n + 1);
+  auto offsetsHost = Kokkos::create_mirror_view(Kokkos::HostSpace(), offsets);
+  Kokkos::Random_XorShift64_Pool<Kokkos::DefaultHostExecutionSpace> gHost(1931);
+  Kokkos::fill_random(offsetsHost, gHost, k);
+  Kokkos::DefaultHostExecutionSpace().fence();
+  offsetsHost(0) = 0;
+  // Prefix-sum
+  for (unsigned i = 1; i <= n; i++) {
+    offsetsHost(i) += offsetsHost(i - 1);
+  }
+  Kokkos::deep_copy(offsets, offsetsHost);
+  unsigned totalLength = offsetsHost(n);
+  keys                 = KeyViewType("Keys", totalLength);
+  Kokkos::Random_XorShift64_Pool<typename KeyViewType::execution_space> g(1931);
+  Kokkos::fill_random(keys, g, minKey, maxKey);
+}
+
+// Same as above, but also generate values array with same length as keys
```

Comment:
Avoid duplicating code.
For instance write 3 functions, one to generate offsets, one for keys, and one for values.